packages/assets-controllers/src/TokensController.ts

```diff
@@ -730,28 +730,110 @@ export class TokensController extends BaseControllerV1<
+      );
+    }
+
+    const provider = this._getProvider(networkClientId);
+    const isErc1155 = await safelyExecute(() =>
+      new ERC1155Standard(provider).contractSupportsBase1155Interface(
+        asset.address,
+      ),
+    );
+    if (isErc1155) {
+      throw rpcErrors.invalidParams(
+        `Contract ${asset.address} must match type ${type}, but was detected as ${ERC1155}`,
+      );
+    }
+
+    const erc20 = new ERC20Standard(provider);
+    const [contractName, contractSymbol, contractDecimals] = await Promise.all([
+      safelyExecute(() => erc20.getTokenName(asset.address)),
+      safelyExecute(() => erc20.getTokenSymbol(asset.address)),
+      safelyExecute(async () => erc20.getTokenDecimals(asset.address)),
+    ]);
+
+    asset.name = contractName;
+
+    // Validate symbol
+
+    if (!asset.symbol && !contractSymbol) {
+      throw rpcErrors.invalidParams(
+        'A symbol is required, but was not found in the request or contract',
+      );
+    }
+
+    if (
+      contractSymbol !== undefined &&
+      asset.symbol !== undefined &&
+      asset.symbol !== contractSymbol
+    ) {
+      throw rpcErrors.invalidParams(
+        `The symbol in the request (${asset.symbol}) does not match the symbol in the contract (${contractSymbol})`,
+      );
+    }
+
+    asset.symbol ??= contractSymbol;
+    if (typeof asset.symbol !== 'string') {
+      throw rpcErrors.invalidParams(`Invalid symbol: not a string`);
+    }
+
+    if (asset.symbol.length > 11) {
+      throw rpcErrors.invalidParams(
+        `Invalid symbol "${asset.symbol}": longer than 11 characters`,
+      );
+    }
+
+    // Validate decimals
+
+    if (asset.decimals === undefined && contractDecimals === undefined) {
+      throw rpcErrors.invalidParams(
+        'Decimals are required, but were not found in the request or contract',
+      );
+    }
+
+    if (
+      contractDecimals !== undefined &&
+      asset.decimals !== undefined &&
+      String(asset.decimals) !== contractDecimals
+    ) {
+      throw rpcErrors.invalidParams(
+        `The decimals in the request (${asset.decimals}) do not match the decimals in the contract (${contractDecimals})`,
+      );
+    }
+
+    const decimalsStr = asset.decimals ?? contractDecimals;
+    const decimalsNum = parseInt(decimalsStr as unknown as string, 10);
+    if (!Number.isInteger(decimalsNum) || decimalsNum > 36 || decimalsNum < 0) {
+      throw rpcErrors.invalidParams(
+        `Invalid decimals "${decimalsStr}": must be an integer 0 <= 36`,
+      );
+    }
+    asset.decimals = decimalsNum;
 
     const suggestedAssetMeta: SuggestedAssetMeta = {
       asset,
       id: this._generateRandomId(),
       time: Date.now(),
       type,
-      interactingAddress: interactingAddress || selectedAddress,
+      interactingAddress: interactingAddress || this.config.selectedAddress,
     };
 
-    validateTokenToWatch(asset);
```

Comment:
[nit] doesn't really matter but could have preserved this utility function and just passed in the provider do those operations.

Reply:
Yeah it could have. Just didn't feel like it gained much from being separate.  Already tightly coupled to `watchAsset`, not more testable, and would need to go beyond validation by returning data to avoid duplicating logic.